File: Napakalaki/src/NapakalakiGame/Napakalaki.java
/**
 * Napakalaki.java
 */

package NapakalakiGame;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * Representa una partida.
 * Singleton.
 */
public class Napakalaki {
    
    private static Napakalaki instance = null;
    
    CardDealer dealer;
    Monster currentMonster;
    Player currentPlayer;
    ArrayList<Player> players;
    
    /**
     * Constructor
     */
    private Napakalaki() {
        dealer = CardDealer.getInstance();
        currentMonster = null;
        currentPlayer = null;
        players = new ArrayList();
    }
    
    /**
     * @return instancia del singleton
     */
    public static Napakalaki getInstance() {
        if (instance == null)
            instance = new Napakalaki();
        
        return instance;
    }
    
    /**
     * Inicializa los jugadores.
     * @param names Lista de nombres
     * @pre 2 <= names.size() <= 3
     */
    private void initPlayers(ArrayList<String> names) {
        for (String n : names)
            players.add(new Player(n));
    }

    private Player nextPlayer() {
        int nextP;
        int numPlayers = players.size();
        Random rand = new Random();
        
        if (currentPlayer == null) {
            nextP = rand.nextInt(numPlayers);
        }
        
        else {
            nextP = (players.indexOf(currentPlayer) + 1) % numPlayers;
        }
        
        currentPlayer = players.get(nextP);
        
        return currentPlayer;
    }
    
    private boolean nextTurnAllowed() {
        return currentPlayer == null || currentPlayer.validState();
    }
    
    private void setEnemies() {
        Random rand = new Random();
        
        for (Player p : players) {
            int playerIndex = players.indexOf(p);
            int enemyIndex;
            do {
                enemyIndex = rand.nextInt(players.size());
            } while (enemyIndex == playerIndex);
            
            p.setEnemy(players.get(enemyIndex));
        }
    }
    
    public CombatResult developCombat() {
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        
        dealer.giveMonsterBack(currentMonster);
        
        if (combatResult == CombatResult.LOSEANDCONVERT) {
            Cultist cultist = dealer.nextCultist();
            CultistPlayer cultistPlayer = new CultistPlayer(currentPlayer, cultist);
            
            // Reemplazar antiguo jugador por el nuevo sectario
            players.set(players.indexOf(currentPlayer), cultistPlayer);
            
            for (Player p : players) {
                if (p.getEnemy() == currentPlayer)
                    p.setEnemy(cultistPlayer);
            }
            currentPlayer = cultistPlayer;
        }
        
        return combatResult;
    }
    
    /**
     * @param treasures 
     * @pre treasures.size() >= 1
     */
    public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
        for (Iterator<Treasure> i = treasures.iterator(); i.hasNext();) {
            Treasure treasure = i.next();
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    /**
     * @param treasures 
     * @pre treasures.size() >= 1
     */
    public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
        for (Iterator<Treasure> i = treasures.iterator(); i.hasNext();) {
            Treasure treasure = i.next();
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    /**
     * @param treasures
     * @pre treasures.size() >= 1
     */
    public void makeTreasuresVisible(ArrayList<Treasure> treasures) {
        for (Treasure t : treasures) {
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> players) {
        initPlayers(players);
        setEnemies();
        dealer.initCards(); 
        nextTurn();
    }

    public Monster getCurrentMonster() {
        return currentMonster;
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }
    
    public boolean nextTurn() {
        boolean stateOK = nextTurnAllowed();
        
        if (stateOK) {
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
        }
        
        if (currentPlayer.isDead()) {
            currentPlayer.initTreasures();
        }
        
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result) {
        return result == CombatResult.WINGAME;
    }
    
    @Override
    public String toString() {
        String text = "Players: ";
        
        /**
         * Aprovechamos los streams de Java 8.
         * Player::getName hace referencia al método de instancia 'getName' de una instancia
         * arbitraria de Player.
         */
        text += players.stream().map(Player::getName).collect(Collectors.joining(", "));
        
        text += "\nCurrent player: ";
        if (currentPlayer != null)
            text += currentPlayer.getName();
        
        text += "\nCurrent monster: ";
        if (currentMonster != null)
            text += currentMonster.getName();
        
        return text;
    }
    
}

File: Napakalaki/lib/Napakalaki.rb
# encoding: utf-8

require_relative 'combat_result'
require_relative 'card_dealer'
require_relative 'player'
require 'singleton'

module NapakalakiGame

  # class Napakalaki: representa una instancia del juego
  class Napakalaki
    include Singleton  # la clase es <<singleton>>

    attr_reader :currentMonster  # monstruo actual
    attr_reader :currentPlayer   # jugador actual

    def initialize
      @currentMonster = nil
      @currentPlayer = nil
      @players = Array.new
      @dealer = CardDealer.instance
    end

    def getCurrentPlayer
      @currentPlayer
    end

    def getCurrentMonster
      @currentMonster
    end

    private
    
    def initPlayers(names)
      names.each { |n| @players << Player.new(n) }
    end

    def nextPlayer
      num_players = @players.length
      if @currentPlayer.nil? 
        nextP = rand(num_players)
      else
        nextP = (@players.index(@currentPlayer) + 1) % num_players
      end
      @currentPlayer = @players[nextP]
    end

    def nextTurnAllowed
      @currentPlayer.nil? or @currentPlayer.validState
    end

    def setEnemies
      @players.each do |p|
        begin
          enemy = rand(@players.length) 
        end while @players[enemy] == p
        p.enemy = @players[enemy]
      end
    end

    public
    
    def developCombat
      combatResult = @currentPlayer.combat(@currentMonster)

      if combatResult == CombatResult::LOSEANDCONVERT
        c = @dealer.nextCultist
        p = CultistPlayer.new(@currentPlayer, c)
        @players.at(@players.index(@currentPlayer)) = p

        @players.each do |player|
          if player.getEnemy == @currentPlayer
            player.enemy = p
          end
        end
        
        @currentPlayer = p
      end
      
      @dealer.giveMonsterBack(@currentMonster)
      return combatResult
    end

    # @pre: treasures no vacío
    def discardVisibleTreasures(treasures)
      treasures.each do |t|
        @currentPlayer.discardVisibleTreasure(t)
        @dealer.giveTreasureBack(t)
      end
    end

    # @pre: treasures no vacío
    def discardHiddenTreasures(treasures)
      treasures.each do |t|
        @currentPlayer.discardHiddenTreasure(t)
        @dealer.giveTreasureBack(t)
      end
    end

    # @pre: treasures no vacío
    def makeTreasuresVisible(treasures)
      treasures.each do |t|
        @currentPlayer.makeTreasureVisible(t)
      end
    end

    def initGame(players)
      initPlayers(players)
      setEnemies
      @dealer.initCards
      nextTurn
    end

    def nextTurn
      stateOK = nextTurnAllowed
      
      if stateOK
        @currentMonster = @dealer.nextMonster
        @currentPlayer = nextPlayer

        if @currentPlayer.isDead
          @currentPlayer.initTreasures
        end
      end
      
      return stateOK
    end

    def endOfGame(result)
      result == CombatResult::WINGAME
    end
    
    def to_s
      text = "Jugadores: " + @players.map(&:name).join(", ") + "\nJugador actual: "
      
      text << "#{@currentPlayer.name}" unless @currentPlayer.nil?
      text << "\nMonstruo actual: "
      text << "#{@currentMonster.name}" unless @currentMonster.nil?
      text
    end

  end # Napakalaki

end # NapakalakiGame
